politeiad/backendv2/tstorebe/plugins/pi/cmds_test.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) 2021 The Decred developers
+// Use of this source code is governed by an ISC license that can be found in
+// the LICENSE file.
+
+package pi
+
+import (
+	"encoding/hex"
+	"encoding/json"
+	"errors"
+	"strconv"
+	"testing"
+
+	"github.com/decred/politeia/politeiad/api/v1/identity"
+	backend "github.com/decred/politeia/politeiad/backendv2"
+	"github.com/decred/politeia/politeiad/plugins/pi"
+)
+
+// billingStatusTest contains the input and output for a test that tests
+// the cmdBillingStatus func.
+type billingStatusTest struct {
+	name  string    // Test name
+	input testInput // Input
+	err   error     // Expected output
+}
+
+type testInput struct {
+	token   []byte
+	payload string
+}
+
+func TestCmdBillingStatus(t *testing.T) {
```

Comment:
One of the primary benefits of using table driven tests is the ability to write clean, easy to read tests that do not duplicate code.  These tests here need some work to make them more readable.  Here are some examples of what I mean.  Convert these existing tests to use this type of format.

```
func TestCmdBillingStatus(t *testing.T) {
	// Setup pi plugin
	p, cleanup := newTestPiPlugin(t)
	defer cleanup()

	// Setup an identity that will be used to create the payload
	// signatures.
	fid, err := identity.New()
	if err != nil {
		t.Fatal(err)
	}

	// Setup test data
	var (
		// Valid input
		token     = "45154fb45664714b"
		status    = pi.BillingStatusCompleted
		publicKey = fid.Public.String()

		msg        = token + strconv.Itoa(int(status))
		signatureb = fid.SignMessage([]byte(msg))
		signature  = hex.EncodeToString(signatureb[:])

		// signatureIsWrong is a valid hex encoded, ed25519 signature,
		// but that does not correspond to the valid input parameters
		// listed above.
		signatureIsWrong = "b387f678e1236ca1784c4bc77912c754c6b122dd8b" +
			"3e499617706dd0bd09167a113e59339d2ce4b3570af37a092ba88f39e7f" +
			"c93a5ac7513e52dca3e5e13f705"
	)
	tokenb, err := hex.DecodeString(token)
	if err != nil {
		t.Fatal(err)
	}

	// Setup tests
	var tests = []struct {
		name  string // Test name
		token []byte
		sbs   pi.SetBillingStatus
		err   error // Expected error output
	}{
		{
			"payload token invalid",
			tokenb,
			setBillingStatus(t, fid,
				pi.SetBillingStatus{
					Token:  "zzz",
					Status: status,
					Reason: "",
				}),
			pluginError(pi.ErrorCodeTokenInvalid),
		},
		{
			"payload token does not match cmd token",
			tokenb,
			setBillingStatus(t, fid,
				pi.SetBillingStatus{
					Token:  "da70d0766348340c",
					Status: status,
					Reason: "",
				}),
			pluginError(pi.ErrorCodeTokenInvalid),
		},
		{
			"signature is not hex",
			tokenb,
			pi.SetBillingStatus{
				Token:     token,
				Status:    status,
				Reason:    "",
				PublicKey: publicKey,
				Signature: "zzz",
			},
			pluginError(pi.ErrorCodeSignatureInvalid),
		},
		{
			"signature is the wrong size",
			tokenb,
			pi.SetBillingStatus{
				Token:     token,
				Status:    status,
				Reason:    "",
				PublicKey: publicKey,
				Signature: "123456",
			},
			pluginError(pi.ErrorCodeSignatureInvalid),
		},
		{
			"signature is wrong",
			tokenb,
			pi.SetBillingStatus{
				Token:     token,
				Status:    status,
				Reason:    "",
				PublicKey: publicKey,
				Signature: signatureIsWrong,
			},
			pluginError(pi.ErrorCodeSignatureInvalid),
		},
	}

	// Run tests
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			// Setup command payload
			b, err := json.Marshal(tc.sbs)
			if err != nil {
				t.Fatal(err)
			}
			payload := string(b)

			// Run test
			_, err = p.cmdBillingStatus(tc.token, payload)
                        // TODO: this is not complete, it has been abbreviated for clarity.  Fill back in.
		})
	}
}

// setBillingStatus uses the provided arguments to return a SetBillingStatus
// with a valid PublicKey and Signature.
func setBillingStatus(t *testing.T, fid *identity.FullIdentity, sbs pi.SetBillingStatus) pi.SetBillingStatus {
	t.Helper()

	msg := sbs.Token + strconv.Itoa(int(sbs.Status)) + sbs.Reason
	sig := fid.SignMessage([]byte(msg))

	return pi.SetBillingStatus{
		Token:     sbs.Token,
		Status:    sbs.Status,
		Reason:    sbs.Reason,
		PublicKey: fid.Public.String(),
		Signature: hex.EncodeToString(sig[:]),
	}
}

// pluginError returns a backend PluginError for the provided pi ErrorCodeT.
func pluginError(e pi.ErrorCodeT) error {
	return backend.PluginError{
		PluginID:  pi.PluginID,
		ErrorCode: uint32(e),
	}
}
```